vectara_agentic/tools.py

```diff
@@ -186,30 +190,60 @@ def create_rag_tool(
                 Each dictionary should specify the "type" of reranker (mmr, slingshot, udf)
                 and any other parameters (e.g. "limit" or "cutoff" for any type,
                 "diversity_bias" for mmr, and "user_function" for udf).
-            If using slingshot/multilingual_reranker_v1, it must be first in the list.
+                If using slingshot/multilingual_reranker_v1, it must be first in the list.
             include_citations (bool, optional): Whether to include citations in the response.
                 If True, uses markdown vectara citations that requires the Vectara scale plan.
             fcs_threshold (float, optional): a threshold for factual consistency.
                 If set above 0, the tool notifies the calling agent that it "cannot respond" if FCS is too low.
+            verbose (bool, optional): Whether to print verbose output.
 
         Returns:
             VectaraTool: A VectaraTool object.
         """
+
         vectara = VectaraIndex(
             vectara_api_key=self.vectara_api_key,
             vectara_customer_id=self.vectara_customer_id,
             vectara_corpus_id=self.vectara_corpus_id,
             x_source_str="vectara-agentic",
         )
 
-        def _build_filter_string(kwargs):
+        def _build_filter_string(kwargs: Dict[str, Any], tool_args_type: Dict[str, str]) -> str:
             filter_parts = []
+            comparison_operators = [">=", "<=", ">", "<", "!="]
+
             for key, value in kwargs.items():
-                if value:
-                    if isinstance(value, str):
-                        filter_parts.append(f"doc.{key}='{value}'")
+                if value is None or value == "":
+                    continue
+
+                # Determine the prefix for the key. Valid values are "doc" or "part"
+                # default to 'doc' if not specified
+                prefix = tool_args_type.get(key, "doc")
+
+                # Check if value contains a known comparison operator at the start
+                val_str = str(value).strip()
+                matched_operator = None
+                for op in comparison_operators:
+                    if val_str.startswith(op):
+                        matched_operator = op
+                        break
+
+                # Break down operator from value
+                # e.g. val_str = ">2022" --> operator = ">", rhs = "2022"
+                if matched_operator:
+                    rhs = val_str[len(matched_operator):].strip()
+                    if rhs.isdigit() or is_float(rhs):
+                        filter_parts.append(f"{prefix}.{key}{matched_operator}{rhs}")
+                    else:
```

Comment:
It looks like this would prevent the "!=" operator (and the "=" if that is added to comparison_operators as per my earlier comment) from being used for numerical operands. But it is a valid scenario to try to compare an integer-typed filter attribute to an exact integer (e.g. year = 2024). So I think you should change this condition to allow for != and = operators on ints/floats. Or perhaps allow for ints, but continue to disallow for floats.

Another check you can make is to disallow any operator except for != and = if the rhs is a string. I.e. don't allow <, <=, >, and >= on string operands.